### PR TITLE
Java: Add OpenApi plugin importer and upgrade tool calling to cover more usecases

### PR DESCRIPTION
**Pull request by johnoliver**
### Contribution Checklist

- [ ] The code builds clean without any errors or warnings
- [ ] The PR follows the [SK Contribution Guidelines](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md) and the [pre-submission formatting script](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md#development-scripts) raises no violations
- [ ] All unit tests pass, and I have added new tests where possible
- [ ] I didn't break anyone :smile:

### CODE REVIEW

File: java/semantickernel-api/src/main/java/com/microsoft/semantickernel/hooks/KernelHooks.java [hunk 1]
```diff
@@ -102,6 +103,17 @@ public String addPreChatCompletionHook(
...
+     * Add a {@link PreToolCallHook} to the collection of hooks.
+     *
+     * @param function the function to add
+     * @return the key of the hook in the collection
+     */
+    public String addPreToolCallHook(
+        Function<PreToolCallEvent, PreToolCallEvent> function) {
+        return addHook((PreToolCallHook) function::apply);
+    }
+
```
Comment (dsgrieve) [inline java/semantickernel-api/src/main/java/com/microsoft/semantickernel/hooks/KernelHooks.java:116]: No PostToolCallHook? Maybe I have a function that computes degrees Celsius only and I want to convert to Farenheit.

File: java/semantickernel-api/src/main/java/com/microsoft/semantickernel/plugin/KernelPluginFactory.java [hunk 2]
```diff
@@ -89,7 +89,34 @@ public class KernelPluginFactory {
...
+    public static KernelPlugin createFromObject(Class<?> clazz, Object target, String pluginName) {
+        if (!clazz.isInstance(target)) {
+            throw new SKException("Target object is not an instance of the provided class");
+        }
+
+        if (clazz.getName().endsWith("_ClientProxy")) {
+            LOGGER.warn(
+                "The class " + clazz.getName() + " is likely a proxy from a DI framework, "
+                    + "this likely will prevent Semantic Kernel from detecting methods. "
+                    + "Try using: KernelPluginFactory.createFromObject(Class<?> clazz, Object target, String pluginName).");
```
Comment (dsgrieve) [inline java/semantickernel-api/src/main/java/com/microsoft/semantickernel/plugin/KernelPluginFactory.java:116]: Aren't you telling me to use the same function I just called?

File: java/semantickernel-api/src/main/java/com/microsoft/semantickernel/plugin/KernelPluginFactory.java [hunk 3]
```diff
@@ -89,7 +89,34 @@ public class KernelPluginFactory {
...
+     *                   value is {@code null}, a plugin name is derived from the type of the
+     *                   target.
+     * @return The new plugin.
+     */
+    public static KernelPlugin createFromObject(Class<?> clazz, Object target, String pluginName) {
+        if (!clazz.isInstance(target)) {
+            throw new SKException("Target object is not an instance of the provided class");
+        }
+
+        if (clazz.getName().endsWith("_ClientProxy")) {
```
Comment (dsgrieve) [inline java/semantickernel-api/src/main/java/com/microsoft/semantickernel/plugin/KernelPluginFactory.java:112]: Isn't this dependent on the DI framework? Could this work as, if reflection fails then warn that a probable cause is proxy? Another possible reason would be a module that doesn't export or open.

File: java/semantickernel-api/src/main/java/com/microsoft/semantickernel/plugin/KernelPluginFactory.java (hunk 2 above)
Comment (johnoliver) [reply java/semantickernel-api/src/main/java/com/microsoft/semantickernel/plugin/KernelPluginFactory.java:116]: Not necessarily, they may have got to this position through the invocation without the Class argument

File: java/semantickernel-api/src/main/java/com/microsoft/semantickernel/hooks/KernelHooks.java (hunk 1 above)
Comment (johnoliver) [reply java/semantickernel-api/src/main/java/com/microsoft/semantickernel/hooks/KernelHooks.java:116]: At the moment I did not have a usecase for PostToolCall hook, @milderhc is looking at the filtering requirement, I would leave it to him to see if it is needed for the filtering work

File: java/semantickernel-api/src/main/java/com/microsoft/semantickernel/plugin/KernelPluginFactory.java (hunk 3 above)
Comment (johnoliver) [reply java/semantickernel-api/src/main/java/com/microsoft/semantickernel/plugin/KernelPluginFactory.java:112]: It is framework specific, I will also add a warning in the case of no methods found

File: java/semantickernel-api/src/main/java/com/microsoft/semantickernel/hooks/KernelHooks.java (hunk 1 above)
Comment (dsgrieve) [reply java/semantickernel-api/src/main/java/com/microsoft/semantickernel/hooks/KernelHooks.java:116]: Odd to have pre/post for other events and not this one.

**Pull request merged**